Complete project refactoring || Merge offer

max-koro opened the pull request:
Hey @ramaureirac !
Thank you for your patience :D

<hr>

### New features:
- Docs:
    - Complete native Godot editor documentation across the project
![native_doc](https://github.com/user-attachments/assets/47ee9852-2408-4879-8975-220fb266120a)

- Debug:
    - Toggleable Debugger
        - Sending to Output tab
        - Disable in TacticsConfig
![debugger](https://github.com/user-attachments/assets/8098806a-92f6-4ee4-9450-b598093df7d7)

- Gameplay:
    - Automatic selection of the first player pawn on opponent turn's ending
    - When the player needs to select a target, all opponent pawns have their HP displayed above them to facilitate decision-making

- Camera:
    - Added a "free Look" camera mode (try holding MMB)
        - Upon releasing MMB, the camera automatically snaps to the nearest quadrant
    - Added a camera zoom using scroll wheel
    - Added "screen border hover" cursor panning
    - Added global map border panning limits (to avoid accidental extreme panning in windowed mode)
        - Added an export value under TacticsLevel that locally overrides said radius, for particularly large maps
    - Removed camera focus shifting on pawn hover
        - Sub-optimal game design imho -- accidental hover happens easily + players usually want to read stats on hover, not move camera
        - Instead, focus happens on click

- Hints:
    - Collapsible input hints
    - During player target selection, all enemy pawns health is displayed

### Project refactoring:
[View Structure Overview Diagram](https://github.com/user-attachments/assets/65bb6862-6e84-4149-af5a-047a04f413eb)
- Made project structure more scalable (shared assets architecture)
    - Models: Centralized storage for class parameters & logic services
    - Modules: Self-contained reusable units, they connect to their models as needed
    - Dedicated maps directory under `assets`
    - Other smaller architecture changes to nest components logically

- Optimized framework performance:
    - static typing
    - pattern efficiency
    - iteration-saving
    - etc.

- Adapted code to follow [the official GDScript style guidelines](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#one-statement-per-line), improving readability

- Leveraged Godot 4 latest features (that I know of) where it made sense

<hr>

If you want me to make my own fork, I'll do that 😄 
Otherwise I'd be happy to become co-contributor and help you maintain the repo!

Feel free to use the native GitHub review system to leave comments 🙂 
All criticism welcome!

mbusson committed:
TacticsPawn + TacticsPawnService ground work

mbusson committed:
TacticsPlayer: halfway there

mbusson committed:
Small fix

mbusson committed:
TacticsPlayer: ground work done

mbusson committed:
Minor linting

mbusson committed:
TacticsControls ground work

mbusson committed:
Update tactics_controls.gd

mbusson committed:
Tactics Opponent part 1

mbusson committed:
TacticsOpponent ground work + small refactors

mbusson committed:
Bunch of refactors

mbusson committed:
Small refactor

mbusson committed:
Entity restructuring pt. 1

mbusson committed:
Restructuring

mbusson committed:
Final structure adjustments + actor Stats class

mbusson committed:
Hero profession Resources + migration prep

mbusson committed:
Mobs resources + finished resource mapping

mbusson committed:
Successfully migrated pawn stats

mbusson committed:
Small bugfix

mbusson committed:
TacticsPawn refactor: 1/2

mbusson committed:
TacticsPawn refactor: 2/2

mbusson committed:
Minor changes to PC input label & method names

mbusson committed:
TacticsParticipant: done.

mbusson committed:
Unfound nodes fix

mbusson committed:
TacticsOpponent: done.

mbusson committed:
Save before big camera behaviour changes

mbusson committed:
Import refactor

mbusson committed:
Reinstated base stats

mbusson committed:
Update README.md

mbusson committed:
Fixed turn-handling edge-case + minor adjustments

mbusson committed:
Removed Addons for MR

mbusson committed:
Static Typing fixes

max-koro commented on LICENSE:
Accidentally deleted -- please reinstate :)

mbusson commented on assets/sprites/labels/controls-ui-xbox.png:
File moved to `assets/textures`

max-koro commented on docs/.gdignore:
Accidentally deleted the 'docs' folder -- please reinstate :)

max-koro reviewed (commented):
Markers for files I accidentally removed during the transition

mbusson committed:
Update README.md

ramaureirac commented on assets/scene/mai1ACA.tmp:
nit: is this a temporary file? if so should we delete it?

max-koro replied:
I believe we can! Never ran into any issues after deleting tmp files thus far

ramaureirac commented on data/models/world/combat/participant/pawn/service/combat.gd:
```diff
@@ -0,0 +1,35 @@
+## @param target_pawn: The TacticsPawn being attacked
+## @param delta: Time elapsed since the last frame
+## @return: Whether the attack was completed
+func attack_target_pawn(pawn: TacticsPawn, target_pawn: TacticsPawn, delta: float) -> bool:
+	# Make the attacking pawn face the target
+	pawn.serv.movement.look_at_direction(pawn, target_pawn.global_position - pawn.global_position)
+	
+	# Check if the pawn can attack and enough time has passed for the attack animation
+	if pawn.res.can_attack and pawn.res.wait_delay > TacticsPawnResource.MIN_TIME_FOR_ATTACK / 4.0:
+		# Apply damage to the target pawn
+		target_pawn.stats.apply_to_curr_health(-pawn.stats.attack_power)
+		
+		# Print debug information if debug mode is enabled
+		if DebugLog.debug_enabled:
+			print_rich("[color=pink]Attacked ", target_pawn, " for ", pawn.stats.attack_power, " damage.[/color]")
+		
+		# Set the attacking state to false
+		pawn.res.set_attacking(false)
+	
+	# If the minimum time for attack hasn't passed, increment the wait delay
+	if pawn.res.wait_delay < TacticsPawnResource.MIN_TIME_FOR_ATTACK:
+		pawn.res.wait_delay += delta
+		return false
+	
+	# Reset the wait delay and return true to indicate the attack is complete
+	pawn.res.wait_delay = 0.0
+	return true
```
I think there is an issue when performing an attack.

For some reason the health stats are not being displayed during the attack. because of this its kinda hard to get a feedback for the player since this value should change in realtime.

I guess adding a "target_pawn.show_pawn_stats(true)" at the beginning should work. but not sure if here is the correct place.

max-koro replied:
Good point! This can be fixed with the following edit in `res://data/models/world/combat/participant/service/combat.gd`:

```
func attack_pawn(delta: float, is_player: bool) -> void:
	# Handle case when no attackable pawn is available
	if not res.attackable_pawn:
		res.curr_pawn.res.can_attack = false
	else:
		# Attempt to attack the target pawn
		if not res.curr_pawn.attack_target_pawn(res.attackable_pawn, delta):
			return
		# Hide actions menu and focus camera on attacking pawn
		controls.set_actions_menu_visibility(false, res.attackable_pawn)
		camera.target = res.curr_pawn
	
	# Reset attackable pawn
	res.attackable_pawn = null
	# Reset opponent stats display
	if res.display_opponent_stats:
		res.display_opponent_stats = false
	
	# Determine next stage based on current pawn's ability to act and whether it's a player pawn
	if not res.curr_pawn.can_act() or not is_player:
		res.stage = res.STAGE_SELECT_PAWN
	elif res.curr_pawn.can_act() and is_player:
		res.stage = res.STAGE_SHOW_ACTIONS
```

ramaureirac commented on assets/scene/main.tscn:
```diff
@@ -0,0 +1,40 @@
+offset_right = 161.0
+offset_bottom = 307.0
+
+[node name="LoadMap0" type="Button" parent="UI/MapSelector"]
+layout_mode = 0
+offset_left = 517.0
+offset_top = 302.0
+offset_right = 643.0
+offset_bottom = 362.0
+text = "Map 1"
+icon = SubResource("ImageTexture_sa2p4")
```
nit: (quite sure i'm the only one affected by this haha) but for some reason when i switch to fullscreen this button stops working. i've been testing this on a macbook (probably that's why)

ramaureirac replied:
also, the button got decentralized

max-koro replied:
Replicated! (And fixed)

It was a Node config issue, thanks for catching it!
I'll submit this fix in an upcoming merge :D (Hard to edit that one in git)

ramaureirac reviewed (changes requested):
LGTM! left few comments.
Pls review before merging
Ty!

mbusson committed:
Merge branch 'development' into development

mbusson committed:
Update combat.gd

https://github.com/ramaureirac/godot-tactical-rpg/pull/16/#discussion_r1721618806

mbusson committed:
Delete assets/scene/mai1ACA.tmp

mbusson committed:
Update combat.gd

Edited wrong file initially -- quick fix

mbusson committed:
All good but borked free_look (again)

mbusson committed:
Merge branch 'development' of https://github.com/mbusson/godot-tactical-rpg-3d into development

mbusson committed:
All fixed

mbusson committed:
Adding missing files

mbusson committed:
Made Windowed by default
